src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -0,0 +1,173 @@
+// <copyright file="Security.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Datadog.Trace.AppSec.Agent;
+using Datadog.Trace.AppSec.EventModel;
+using Datadog.Trace.AppSec.Transport;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec
+{
+    /// <summary>
+    /// The Secure is responsible cooridating app sec
+    /// </summary>
+    public class Security : IDatadogSecurity
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<Security>();
+
+        private static Security _instance;
+        private static bool _globalInstanceInitialized;
+        private static object _globalInstanceLock = new();
+
+        private readonly IPowerWaf _powerWaf;
+        private readonly IAgentWriter _agentWriter;
+        private readonly InstrumentationGateway _instrumentationGateway;
+        private readonly ConcurrentDictionary<Guid, Action> toExecute = new ConcurrentDictionary<Guid, Action>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Security"/> class with default settings.
+        /// </summary>
+        public Security()
+            : this(null, null)
+        {
+        }
+
+        internal Security(InstrumentationGateway instrumentationGateway = null, IPowerWaf powerWaf = null, IAgentWriter agentWriter = null)
+        {
+            try
+            {
+                var enabled = Environment.GetEnvironmentVariable(ConfigurationKeys.AppSecEnabled)?.ToBoolean();
+                Enabled = enabled == true;
+
+                var blockingEnabled = Environment.GetEnvironmentVariable(ConfigurationKeys.AppSecBlockingEnabled)?.ToBoolean();
+                BlockingEnabled = blockingEnabled == true;
+
+                Log.Information($"Security.Enabled: {Enabled}, Security.BlockingEnabled {BlockingEnabled}");
+
+                _instrumentationGateway = instrumentationGateway ?? new InstrumentationGateway();
+
+                _powerWaf = powerWaf ?? (Enabled ? new PowerWaf() : new NullPowerWaf());
+                _agentWriter = agentWriter ?? new AgentWriter();
```

Comment:
The AgentWriter uses background threads. Maybe we should refrain from initializing it if appsec isn't enabled.

Reply:
Fixed by ca5f5ac9bd94067865032176d9db26d1a3e2c17c